Integrations/python/deephaven/learn/__init__.py

```diff
@@ -0,0 +1,129 @@
+        _Output_ = jpy.get_type("io.deephaven.integrations.learn.Output")
+        _Computer_ = jpy.get_type("io.deephaven.integrations.learn.Computer")
+        _Scatterer_ = jpy.get_type("io.deephaven.integrations.learn.Scatterer")
+
+
+# every module method that invokes Java classes should be decorated with @_passThrough
+@wrapt.decorator
+def _passThrough(wrapped, instance, args, kwargs):
+    """
+    For decoration of module methods, to define necessary symbols at runtime
+
+    :param wrapped: the method to be decorated
+    :param instance: the object to which the wrapped function was bound when it was called
+    :param args: the argument list for `wrapped`
+    :param kwargs: the keyword argument dictionary for `wrapped`
+    :return: the decorated version of the method
+    """
+
+    _defineSymbols()
+    return wrapped(*args, **kwargs)
+
+
+@_passThrough
+class Input:
+    """
+    Provides an interface for getting data from Deephaven tables to Python objects that popular deep learning libraries
+    know how to use. Avoids many inefficencies of Python and allows for seamless integration of Deephaven's real-time
+    capabilities with modern AI frameworks.
+    """
+    def __init__(self, colNames, gatherFunc):
+        self.input = _Input_(colNames, gatherFunc)
+
+
+@_passThrough
+class Output:
+    """
+    Provides an interface for getting data from Python objects like Tensorflow Tensors into Deephaven tables. Enables
+    simple but powerful user-configurability via the scatter function, and works with Deephaven's native real-time capabilities.
+    """
+    def __init__(self, colName, scatterFunc, type=None):
+        self.output = _Output_(colName, scatterFunc, type)
+
+
+@_passThrough
+def eval(table=None, model_func=None, inputs=[], outputs=[], batch_size = None):
+    """
+    Takes relevant data from Deephaven table using inputs, converts that data to the desired Python type, feeds
+    it to model_func, and stores that output in a Deephaven table using outputs.
+
+    :param table: the Deephaven table to perform computations on
+    :param model_func: the function that takes in data and performs AI computations. NOTE that model_func must have
+                       'target' and 'features' arguments in that order
+    :param inputs: the list of Input objects that determine which columns get extracted from the Deephaven table
+    :param outputs: the list of Output objects that determine how to store output from model_func into the Deephaven table
+    """
+
+    # reminder to do error checking on the set of inputs and the set of outputs
+
+    if batch_size == None:
+        if table.isLive():
+            raise ValueError("Batch size cannot be inferred on a live table. Please specify a batch size.")
+        batch_size = table.size()
+
+    computer = _Computer_(table, model_func, [input.input for input in inputs], batch_size)
+    QueryScope.addParam("computer", computer)
+
```

Comment:
I've looked into the memory management.  I think we are holding more memory than is necessary.

If I were going to handle the memory management, I would:
1) Modify `Computer` to (a) keep track of all new `FutureOffsets` created since `clear` was last called and (b) call `FutureOffset.clear()` clear on all tracked `FutureOffsets` when `Computer.clear()` is called.
3) Modify the queries in `exec` to move `__compute.clear()` so that it is just before `dropColumns`.

I think this combination of actions will release as much memory as possible while minimizing the number of calls.